src/nvim/statusline.c

```diff
@@ -342,11 +338,20 @@ void win_redr_ruler(win_T *wp, bool always)
         }
       }
 
-      ScreenGrid *grid = part_of_status ? &default_grid : &msg_grid_adj;
+      ScreenGrid *grid = part_of_status || p_ch <= 0 ? &default_grid : &msg_grid_adj;
       grid_puts(grid, buffer, row, this_ru_col + off, attr);
       grid_fill(grid, row, row + 1,
                 this_ru_col + off + (int)strlen(buffer), off + width, fillchar,
                 fillchar, attr);
+      if (p_ch <= 0) {
+        // UPD_CLEAR is needed for ruler with cmdheight=0
+        must_redraw = UPD_CLEAR;
```

Comment:
still using UPD_CLEAR. This is not going to be a good experience, so what's the point?

Reply:
I have tested other redraws.  But it does not work well.
If other method works and better, please tell me.